# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/localstorage/ActorsParent.cpp
# Commit: dca9f407ea90
# Full Hash: dca9f407ea90a83ae23be6c48aab8ce3f0f6e23e
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-06-14 20:49:45
# Regressor Bug: 1883353
# File Overlap Count: 1
# Description:
#   Bug 1883353 - Require to always drop directory locks explicitly; r=dom-storage-reviewers,jari
#   
#   Until now, directory locks were dropped when the last strong reference was
#   removed or after calling Drop explicitly. The dependency on ref-counting makes
#   it less obvious when directory locks are dropped for real and it's also
# ==============================================================================

diff -r a82c4a895b9d -r dca9f407ea90 dom/localstorage/ActorsParent.cpp
--- a/dom/localstorage/ActorsParent.cpp	Wed Jun 12 15:12:34 2024 +0000
+++ b/dom/localstorage/ActorsParent.cpp	Fri Jun 14 07:21:24 2024 +0000
@@ -77,6 +77,7 @@
 #include "mozilla/dom/quota/Client.h"
 #include "mozilla/dom/quota/ClientImpl.h"
 #include "mozilla/dom/quota/DirectoryLock.h"
+#include "mozilla/dom/quota/DirectoryLockInlines.h"
 #include "mozilla/dom/quota/FirstInitializationAttemptsImpl.h"
 #include "mozilla/dom/quota/OriginScope.h"
 #include "mozilla/dom/quota/PersistenceType.h"
@@ -4414,7 +4415,7 @@
     // There's no connection, so it's safe to release the directory lock and
     // unregister itself from the hashtable.
 
-    mDirectoryLock = nullptr;
+    DropDirectoryLock(mDirectoryLock);
 
     CleanupMetadata();
   }
@@ -5197,7 +5198,7 @@
   // Now it's safe to release the directory lock and unregister itself from
   // the hashtable.
 
-  mDirectoryLock = nullptr;
+  DropDirectoryLock(mDirectoryLock);
 
   CleanupMetadata();
 
@@ -7538,7 +7539,7 @@
     // There's no connection, so it's safe to release the directory lock and
     // unregister itself from the array.
 
-    mDirectoryLock = nullptr;
+    SafeDropDirectoryLock(mDirectoryLock);
 
     CleanupMetadata();
   }
@@ -7551,7 +7552,8 @@
   MOZ_ASSERT(mConnection);
 
   mConnection = nullptr;
-  mDirectoryLock = nullptr;
+
+  DropDirectoryLock(mDirectoryLock);
 
   CleanupMetadata();
 }